airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,170 @@
+
+  private boolean sendCheckpointMessage = false;
+
+  private final DebeziumEventConverter eventConverter;
+
+  public DebeziumMessageProducer(
+                                 final CdcStateHandler cdcStateHandler,
+                                 final CdcTargetPosition targetPosition,
+                                 final DebeziumEventConverter eventConverter,
+                                 final AirbyteFileOffsetBackingStore offsetManager,
+                                 final boolean trackSchemaHistory,
+                                 final AirbyteSchemaHistoryStorage schemaHistoryManager) {
+    this.cdcStateHandler = cdcStateHandler;
+    this.targetPosition = targetPosition;
+    this.eventConverter = eventConverter;
+    this.offsetManager = offsetManager;
+    this.trackSchemaHistory = trackSchemaHistory;
+    this.schemaHistoryManager = schemaHistoryManager;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    this.initialOffset = new HashMap<>(this.previousCheckpointOffset);
+    resetCheckpointValues();
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(ConfiguredAirbyteStream stream) {
+    LOGGER.info("Sending CDC checkpoint state message.");
+    final AirbyteStateMessage stateMessage = createStateMessage(checkpointOffsetToSend);
+    previousCheckpointOffset.clear();
+    previousCheckpointOffset.putAll(checkpointOffsetToSend);
+    sendCheckpointMessage = false;
+    return stateMessage;
+  }
+
+  /**
+   * @param stream
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(ConfiguredAirbyteStream stream, ChangeEventWithMetadata message) {
+
+    if (checkpointOffsetToSend.isEmpty()) {
+      try {
+        final HashMap<String, String> temporalOffset = (HashMap<String, String>) offsetManager.read();
+        if (!targetPosition.isSameOffset(previousCheckpointOffset, temporalOffset)) {
+          checkpointOffsetToSend.putAll(temporalOffset);
+        }
+      } catch (final ConnectException e) {
+        LOGGER.warn("Offset file is being written by Debezium. Skipping CDC checkpoint in this loop.");
+      }
+    }
+
+    if (checkpointOffsetToSend.size() == 1
+        && !message.isSnapshotEvent()) {
+      if (targetPosition.isEventAheadOffset(checkpointOffsetToSend, message)) {
+        sendCheckpointMessage = true;
+      } else {
+        LOGGER.info("Encountered records with the same event offset.");
+      }
+    }
+
+    return eventConverter.toAirbyteMessage(message);
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(ConfiguredAirbyteStream stream) {
+    final var syncFinishedOffset = (HashMap<String, String>) offsetManager.read();
+
+    System.out.println("initialoffset: " + initialOffset);
+    System.out.println("syncFinishedOffset: " + syncFinishedOffset);
```

Comment:
Use the logger, don't print to stdout directly

Reply:
forgot to clean up - was using this to debug a test. fixed.